### PR TITLE
adding SafeIOException

### PR DESCRIPTION
**Pull request by yifeih**
following from: https://github.com/palantir/http-remoting/issues/702

Should this be in preconditions? I don't think it has anything to do with preconditions... Or maybe in safe-logging under the same package?

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/safe-logging](https://github.com/palantir/safe-logging), @yifeih! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeIOException.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+package com.palantir.logsafe.exceptions;
+
+import com.palantir.logsafe.Arg;
+import com.palantir.logsafe.SafeLoggable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+public class SafeIOException extends IOException implements SafeLoggable {
```
Comment (uschi2000) [inline preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeIOException.java]: should be `SafeIoException`, see https://github.com/palantir/gradle-baseline/tree/develop/docs/java-style-guide#camel-case-defined

File: preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeIOException.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    private final List<Arg<?>> arguments;
+
+    public SafeIOException() {
```
Comment (uschi2000) [inline preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeIOException.java]: remove this constructor?

File: preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeIoException.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        super(message);
+        this.arguments = Collections.unmodifiableList(Arrays.asList(arguments));
+    }
+
+    public SafeIoException(String message, Throwable cause, Arg<?>... arguments) {
+        super(message, cause);
+        this.arguments = Collections.unmodifiableList(Arrays.asList(arguments));
+    }
+
+    public SafeIoException(Throwable cause, Arg<?> arguments) {
```
Comment (gracew) [inline preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeIoException.java]: let's just delete this constructor, I don't think it really makes sense for this to take in `args` if there's no message -- the args provide context to the message

**Review by gracew — COMMENTED** [review]
can you also add a corresponding method in  `com.palantir.logsafe.testing.Assertions`?

looks good otherwise!

**Comment by gracew** [comment]
agree this doesn't have to do with preconditions, but let's keep this with the other `SafeXXExceptions` (so current location is good).

**Review by gracew — APPROVED** [review]

**Pull request merged**